table.go

```diff
@@ -1339,63 +1339,121 @@ func (t *TableBlock) abortCompaction(granule *Granule) {
+*/
+func (t *TableBlock) Serialize(writer io.Writer) error {
+	dynCols := map[string][]string{} // NOTE: it would be nice if we didn't have to go back and find all the dynamic columns...
+	t.Index().Ascend(func(i btree.Item) bool {
+		g := i.(*Granule)
 
-	errg := &errgroup.Group{}
-	errg.Go(func() error {
-		for rg := range rowGroupsChan {
-			switch p := rg.(type) {
-			case arrow.Record:
-				b := &bytes.Buffer{}
-				w, err := t.table.config.schema.GetWriter(b, pqarrow.RecordDynamicCols(p))
-				if err != nil {
-					return err
-				}
+		g.PartsForTx(math.MaxUint64, func(p *parts.Part) bool {
+			buf, err := p.AsSerializedBuffer(t.table.config.schema)
+			if err != nil {
+				panic("fuck")
+			}
 
-				if err := pqarrow.RecordToFile(t.table.config.schema, w.ParquetWriter(), p); err != nil {
-					return err
-				}
+			for key, vals := range buf.DynamicColumns() {
+				dynCols[key] = append(dynCols[key], vals...)
+			}
+			return true
+		})
 
-				f, err := parquet.OpenFile(bytes.NewReader(b.Bytes()), int64(b.Len()))
-				if err != nil {
-					return err
-				}
+		return true
+	})
+	dynCols = bufutils.Dedupe(dynCols)
+	w, err := t.table.config.schema.GetWriter(writer, dynCols)
+	if err != nil {
+		return err
+	}
+	defer t.table.config.schema.PutWriter(w)
+	defer w.Close()
 
-				buf, err := dynparquet.NewSerializedBuffer(f)
-				if err != nil {
-					return err
-				}
+	buffSize := 256
+	if t.table.config.rowGroupSize > 0 {
+		buffSize = t.table.config.rowGroupSize
+	}
+	rowGroupRowsWritten := 0
 
-				rowGroups = append(rowGroups, buf.MultiDynamicRowGroup())
+	// Merge all parts in a Granule and write that granule to the file
+	rowsBuf := make([]parquet.Row, buffSize)
+	var ascendErr error
+	t.Index().Ascend(func(i btree.Item) bool {
+		g := i.(*Granule)
 
-			case dynparquet.DynamicRowGroup:
-				rowGroups = append(rowGroups, p)
-			default:
-				return fmt.Errorf("unknown part type: %T", p)
+		// Collect all row groups in each part in granule
+		rgs := []dynparquet.DynamicRowGroup{}
+		g.PartsForTx(math.MaxUint64, func(p *parts.Part) bool {
+			buf, err := p.AsSerializedBuffer(t.table.config.schema)
+			if err != nil {
+				ascendErr = err
+				return false
 			}
+
+			f := buf.ParquetFile()
+			for i := range f.RowGroups() {
+				rgs = append(rgs, buf.DynamicRowGroup(i))
+			}
+
+			return true
+		})
+		if ascendErr != nil {
+			return false
 		}
-		return nil
-	})
 
-	// Collect all the row groups just to determine the dynamic cols
-	err := t.RowGroupIterator(ctx, math.MaxUint64, &AlwaysTrueFilter{}, rowGroupsChan)
-	if err != nil {
-		return err
-	}
+		// Merge the Granule row groups
+		merged, err := t.table.config.schema.MergeRowGroups(dynCols, rgs)
+		if err != nil {
+			ascendErr = err
+			return false
+		}
+		rows := merged.Rows()
+		defer rows.Close()
 
-	close(rowGroupsChan)
-	if err := errg.Wait(); err != nil {
-		return err
+		// Write the merged row groups to the writer
+		for i, n := 0, 0; i < int(merged.NumRows()); i += n {
```

Comment:
Can you reuse the `writeRows` method here?

Reply:
I think this will be a larger rewrite since that's meant to be a one-shot function. But I'll look into what it would take to reuse all that code

Reply:
Definitely let me know if you run into any obstacles. Compaction uses `writeRows` and the only thing I can see that is different is that you don't want to close the writer after writing the merged row group. Maybe you can extract the non-writer logic in `writeRows` to force the caller provide a writer and have the normal path close the writer after one call, while this path uses the same writer.

Reply:
Yea, but it also needs to keep state around rows written for flushing etc. So it needs to be able to handle that state

Reply:
Refactored into a stateful row writer. Let me know what you think